tcl: add Tcl packaging support

agbrooks opened the pull request:
##### Motivation for this change

I am a software engineer at FlightAware, and recently received approval to release some of the tooling that our internal overlay adds for packaging Tcl projects. I think this PR has the potential to make Nix a very compelling tool for working with Tcl (particularly because Tcl does not currently have any actively-maintained package manager).

Currently, it's a bit of a hassle to package software in nixpkgs that depends on Tcl packages. In general, packaging a Tcl application requires you to do the following:
* Manually identify all Tcl packages that your program depends on (directly or transitively) and add a wrapper that will set TCLLIBPATH to the appropriate locations in the store
* Add common Tcl-related configuration flags (thankfully, these are generally pretty standard across Tcl packages and are described by the Tcl Extension Architecture)
* Set TCLLIBPATH appropriately in the checkPhase / installCheckPhase to be able to actually find whatever packages were installed in order to run tests
* Explicitly identify wherever any installed Tcl packages were placed in lib/ and expose that in the passthru to aid other packages in setting up their TCLLIBPATH wrappers

This becomes really unwieldy when you're trying to package something where the dependency tree contains a large number of Tcl packages. At FlightAware, many of our internal Tcl packages have this problem, and it's impractically burdensome just to set up the TCLLIBPATH wrappers manually. Furthermore, the dev experience is also fairly poor compared to something like Python (ie, you can't very well just say `nix-shell -p someTclPackage tcl` and expect to be able to `package require whateverPackage` from `tclsh`).


###### Things done
This PR adds some conveniences for Tcl packaging that are loosely styled after those available for Python. Specifically, there are two major components:

1. A `mkTclDerivation` function, exposed in `tcl`'s passthru. This is just a wrapper around `stdenv.mkDerivation` that mostly just adds a few configuration flags, a dependency on Tcl, and the `tclPackageHook` (described below).
2. A `tclPackageHook`, which populates TCLLIBPATH based on the location of any installed `pkgIndex.tcl` files, propagates this to anti-dependencies, and causes any installed binaries to be automatically wrapped with a TCLLIBPATH wrapper.

This lets you do convenient things like
```
$ nix-shell -p tclx tcl --command "tclsh"
% package require Tclx
8.4
```
which currently isn't possible with Tcl packages in nixpkgs.

I've taken the liberty of incorporating the new Tcl packaging code to many of the Tcl packages already present in nixpkgs. This seems to have been pretty successful in simplifying them.


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
tcl: Add tclPackageHook and mkTclDerivation

Implement tclPackageHook, a setup hook that adjusts TCLLIBPATH to include the
paths of any installed Tcl packages, propagates that TCLLIBPATH to
anti-dependencies, and wraps any installed binaries to set their TCLLIBPATH.

Additionally, implement a makePythonPackage-style mkDerivation wrapper to use
reasonable defaults for Tcl packages and use tclPackageHook.

Commit:
tclreadline: simplify package with tcl.mkTclDerivation

Commit:
incrtcl: Simplify package with tcl.mkTclDerivation

Commit:
itktcl: Simplify package with tcl.mkTclDerivation

Commit:
tcllib: Simplify package with tcl.mkTclDerivation

Commit:
tclx: Leverage tcl.mkTclDerivation

Commit:
tix: Simplify package with tcl.mkTclDerivation

Commit:
tcltls: Simplify package with tcl.mkTclDerivation

Commit:
tk: Simplify package with tcl.mkTclDerivation

Commit:
expect: Simplify package with tcl.mkTclDerivation

Commit:
poke: Simplify package with tcl.tclPackageHook

Commit:
remind: Simplify package with tcl.mkTclDerivation

Commit:
fossil: Remove unneeded TCLLIBPATH export

The use of tcl.mkTclDerivation in tcllib ensures that it will
already be present on TCLLIBPATH when tests are run.

Commit:
bwidget: Simplify bwidget package with tcl.mkTclDerivation

Commit:
tcl2048: Simplify package with tcl.mkTclDerivation

Commit:
xconq: Simplify package with tcl.tclPackageHook

Commit:
maintainers: add agbrooks

Commit:
scid: Simplify package with tcl.mkTclDerivation

Commit:
scid-vs-pc: Simplify package with tcl.mkTclDerivation

Commit:
tix: Prevent tcl.mkTclDerivation from altering configureFlags

agbrooks committed:
scid: Correct typo in maintainers

Commit:
Revert "xconq: Simplify package with tcl.tclPackageHook"

This reverts commit 113826a892512eb2e19985310fd7d4a39db4d75e, which
ended up breaking the Tk import at runtime.

Commit:
bwidget: add missing dependency on tk

Commit:
tcl: fix TCLLIBPATH edge cases in tcl package hook

TCLLIBPATH was not actually extended if a package's path was a substring
of a package already on the path. This commit fixes that behavior.

Additionally, '--prefix' is now used in place of '--set', as the latter
was found to break tkremind from the remind package.

SuperSandro2000 commented on pkgs/applications/editors/poke/default.nix:
```diff
@@ -40,7 +40,7 @@ in stdenv.mkDerivation rec {
 
   buildInputs = [ boehmgc readline ]
-  ++ lib.optional guiSupport tk
+  ++ lib.optionals guiSupport [tk tcl.tclPackageHook tcllib]
```
```suggestion
  ++ lib.optionals guiSupport [ tk tcl.tclPackageHook tcllib ]
```

SuperSandro2000 commented on pkgs/development/interpreters/tcl/generic.nix:
```diff
@@ -1,55 +1,69 @@
+  baseInterp =
+    stdenv.mkDerivation {
+      pname = "tcl";
+      inherit version;
+
+      inherit src;
```
```suggestion
      inherit version src;
```

SuperSandro2000 commented on pkgs/development/interpreters/tcl/tcl-package-hook.sh:
```diff
@@ -0,0 +1,74 @@
+# This hook ensures that we do the following in post-fixup:
+# * wrap any installed executables with a wrapper that configures TCLLIBPATH
+# * write a setup hook that extends the TCLLIBPATH of any anti-dependencies
+
+# Add a directory to TCLLIBPATH, provided that it exists
+_addToTclLibPath() {
+    local -r tclPkg="$1"
+    if [ -z "$tclPkg" ]; then
+        return
+    fi
```
Please consistently use bashism or not.

agbrooks replied:
I've switched to using double-square brackets throughout. Does this address your concern, or was there another bashism/non-bashism that you think should be revisited?

SuperSandro2000 commented on pkgs/development/libraries/bwidget/default.nix:
```diff
@@ -1,30 +1,29 @@
   };
 
   dontBuild = true;
+  propagatedBuildInputs = [tk];
```
```suggestion
  propagatedBuildInputs = [ tk ];
```

SuperSandro2000 commented on pkgs/development/libraries/bwidget/default.nix:
```diff
@@ -1,30 +1,29 @@
   installPhase = ''
-    mkdir -p "$out/lib/${passthru.libPrefix}"
-    cp -R *.tcl lang images "$out/lib/${passthru.libPrefix}"
+    mkdir -p "$out/lib/${libPrefix}"
+    cp -R *.tcl lang images "$out/lib/${libPrefix}"
```
```suggestion
    mkdir -p "$out/lib/bwidget${version}"
    cp -R *.tcl lang images "$out/lib/bwidget${version}"
```

SuperSandro2000 commented on pkgs/development/libraries/bwidget/default.nix:
```diff
@@ -1,30 +1,29 @@
 
-stdenv.mkDerivation rec {
-  pname = "bwidget";
+let
   version = "1.9.14";
+  libPrefix = "bwidget${version}";
+in tcl.mkTclDerivation {
+  pname = "bwidget";
+  inherit version;
```
```suggestion
tcl.mkTclDerivation rec {
  pname = "bwidget";
  version = "1.9.14";
```

SuperSandro2000 commented on pkgs/development/libraries/tclx/default.nix:
```diff
@@ -10,13 +10,11 @@ stdenv.mkDerivation rec {
-
-  configureFlags = [ "--with-tcl=${tcl}/lib" "--exec-prefix=\${prefix}" ];
+  # required in order for tclx to properly detect tclx.tcl at runtime
+  postInstall =
+    let tclXPkg = "tclx${version}";
+        tclXLib = "$prefix/lib/${tclXPkg}";
+     in "ln -s ${tclXLib} ${tclXLib}/${tclXPkg}";
```
```suggestion
  postInstall = ''
    ln -s $prefix/lib/${tclXPkg} $prefix/lib/tclx${version}/tclx${version}
  '';
```

agbrooks committed:
poke: revise formatting

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

agbrooks committed:
tcl: combine inherit expressions

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

agbrooks committed:
bwidget: revise formatting

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

agbrooks committed:
bwidget: make installPhase slightly clearer

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

agbrooks committed:
bwidget: replace verbose let/in expression with rec

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

agbrooks committed:
tclx: simplify postInstall phase

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
tcl: avoid readonly local in package hook

If a path with a space were added to TCLLIBPATH, _addToTclLibPath
would attempt to mutate the tclPkg local and fail.

Commit:
tcl: use double square brackets consistently in tcl package hook

Commit:
tclx: fix undefined variable error

7c6f434c commented:
500+ rebuilds can be a lot or not, but when they include both Chromium and LibreOffice, and then some JDKs, I probably agree that this is large enough for `staging`.

Commit:
tcl: update maintainers

When I reached out to vrthra / Rahul Gopinath about this PR, he
indicated that he no longer has the time to serve as tcl maintainer
and suggested that I take on the responsibility.

agbrooks commented:
@7c6f434c Thanks for the suggestion -- I've updated the PR to merge into `staging`.

agbrooks commented on pkgs/development/interpreters/tcl/generic.nix:
```diff
@@ -1,55 +1,67 @@
+        homepage = "https://www.tcl.tk/";
+        license = licenses.tcltk;
+        platforms = platforms.all;
+        maintainers = with maintainers; [ agbrooks ];
```
Just wanted to note that after reaching out to the previous Tcl maintainer, `vrthra`, by e-mail, he indicated that he no longer has the time to serve as Tcl maintainer and suggested that I take over the responsibility. As such, I've removed him as maintainer and added myself.

7c6f434c commented:
Hm, have you force-pushed a rebased branch? I am not completely sure if merging this as-is would mean also merging a piece of `master` into `staging`

agbrooks commented:
> Hm, have you force-pushed a rebased branch? I am not completely sure if merging this as-is would mean also merging a piece of `master` into `staging`

@7c6f434c No, I'd neglected to do so -- just force-pushed it now. Thanks for pointing that out.

SuperSandro2000 commented:
btw this broke except because now tcl is missing in PATH.

7c6f434c commented:
Oops. Thanks @SuperSandro2000 !

willcohen commented:
I'm trying (and struggling) to deploy this in the context of trying to package `Tcl::Tk`. Since this is already using `buildPerlPackage` so the `mkTclDerivation` abstraction doesn't quite apply, I'm not totally sure where I'm supposed to be injecting Tk to make it available. @agbrooks @nixinator 

With this added to `perl-packages`:
```
  Tcl = buildPerlPackage {
    pname = "Tcl";
    version = "1.27";
    src = fetchurl {
      url = "mirror://cpan/authors/id/V/VK/VKON/Tcl-1.27.tar.gz";
      sha256 = "sha256-+DhYd6Sp7Z89OQPS0PfNcPrDzmgyxg9gCmghzuP7WHI=";
    };
    propagatedBuildInputs = [
      pkgs.tcl
    ] ++ lib.optionals stdenv.isDarwin [
      darwin.apple_sdk.frameworks.CoreServices ];
    meta = {
      description = "Tcl extension module for perl";
      license = with lib.licenses; [ artistic1 gpl1Plus ];
    };
  };

  TclTk = buildPerlPackage {
    pname = "Tcl-Tk";
    version = "1.29";
    src = fetchurl {
      url = "mirror://cpan/authors/id/V/VK/VKON/Tcl-Tk-1.29.tar.gz";
      sha256 = "sha256-nVkLmSSUzRo/ty6gF+gezCptmQNKs6BOBz7ySpqdDmM=";
    };
    propagatedBuildInputs = [
      Tcl
    ];
    meta = {
      description = "Extension module for Perl giving access to Tk via the Tcl extension";
      license = with lib.licenses; [ artistic1 gpl1Plus ];
    };
  };
```

Trying to build `perlPackages.TclTk` yields:

```
$ nix-build $NIXPKGS -A perlPackages.TclTk
this derivation will be built:
  /nix/store/wf2l1cmymwqsgdhkrqf14xy6n1a9vvyx-perl5.34.1-Tcl-Tk-1.29.drv
building '/nix/store/wf2l1cmymwqsgdhkrqf14xy6n1a9vvyx-perl5.34.1-Tcl-Tk-1.29.drv'...
unpacking sources
unpacking source archive /nix/store/463c8n6piircfyw8kl087pjiii6yxcx9-Tcl-Tk-1.29.tar.gz
source root is Tcl-Tk-1.29
setting SOURCE_DATE_EPOCH to timestamp 1532877822 of file Tcl-Tk-1.29/MANIFEST
patching sources
configuring
patching ./t/unicode.t...
patching ./Makefile.PL...
patching ./tk-demos/widget...

Your Tcl installation (tclsh) fails to find Tk package.
One of possible reasons is missing file 'pkgIndex.tcl' in ..../tk8.4/
directory; Please check if you can feed 'package require Tk' to tclsh
no configure script, doing nothing
building
no Makefile, doing nothing
running tests
no Makefile or custom checkPhase, doing nothing
installing
install flags: SHELL=/nix/store/y3hpcrb8gnx6b6pb5b6s247rs13i1rgv-bash-5.1-p16/bin/bash pkgconfigdir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/lib/pkgconfig m4datadir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/share/aclocal aclocaldir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/share/aclocal install
make: *** No rule to make target 'install'.  Stop.
error: builder for '/nix/store/wf2l1cmymwqsgdhkrqf14xy6n1a9vvyx-perl5.34.1-Tcl-Tk-1.29.drv' failed with exit code 2;
       last 10 log lines:
       > One of possible reasons is missing file 'pkgIndex.tcl' in ..../tk8.4/
       > directory; Please check if you can feed 'package require Tk' to tclsh
       > no configure script, doing nothing
       > building
       > no Makefile, doing nothing
       > running tests
       > no Makefile or custom checkPhase, doing nothing
       > installing
       > install flags: SHELL=/nix/store/y3hpcrb8gnx6b6pb5b6s247rs13i1rgv-bash-5.1-p16/bin/bash pkgconfigdir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/lib/pkgconfig m4datadir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/share/aclocal aclocaldir=/nix/store/9j3j2xkmfh3br1d1fldn12kwix6s47jk-perl5.34.1-Tcl-Tk-1.29/share/aclocal install
       > make: *** No rule to make target 'install'.  Stop.
```

willcohen commented:
Digging in a little more, the problem appears to be related to the fact that even though `tk` uses the packaging logic described above, when `perl Makefile.pl` is called during the build phase of that perl package, when perl calls out to `tclsh`, THAT binary doesn’t appear to have access to `tk`. The solution may involve modifying the build phase to address that.

agbrooks commented:
@willcohen I'd wager that setting `TCLLIBPATH` to `"${pkgs.tk}/${pkgs.tk.libdir}"` would allow `tclsh` to find it. Unfortunately, I can't easily confirm whether or not that will fix the problem you're seeing: after adding the snippet you provided to `perl-packages`, I'm not able to build `perlPackages.Tcl` successfully on my NixOS box.
